cmd/pod-scaler/storage.go

```diff
@@ -0,0 +1,136 @@
+	return handle.NewReader(interrupts.Context())
+}
+
+func (b *bucketCache) Store(name string) (io.WriteCloser, error) {
+	handle := b.bucket.Object(name)
+	return handle.NewWriter(context.Background()), nil
+}
+
+func (b *bucketCache) LastUpdated(name string) (time.Time, error) {
+	handle := b.bucket.Object(name)
+	attrs, err := handle.Attrs(interrupts.Context())
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not query cache for attributes: %w", err)
+	}
+	return attrs.Updated, nil
+}
+
+type localCache struct {
+	dir string
+}
+
+var _ Cache = &localCache{}
+
+func (l *localCache) Load(name string) (io.ReadCloser, error) {
+	return os.Open(path.Join(l.dir, name))
+}
+
+func (l *localCache) Store(name string) (io.WriteCloser, error) {
+	return os.Create(path.Join(l.dir, name))
+}
+
+func (l *localCache) LastUpdated(name string) (time.Time, error) {
+	info, err := os.Stat(path.Join(l.dir, name))
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not query cache for attributes: %w", err)
+	}
+	return info.ModTime(), nil
+}
+
+// LoadCache loads cached query data from the canonical location in GCS.
+func LoadCache(loader Loader, metricName string, logger *logrus.Entry) (*CachedQuery, error) {
+	logger.Info("Reading Prometheus data from cache.")
+	var cache CachedQuery
+	reader, err := loader.Load(metricName + ".json")
+	if err != nil {
+		return nil, fmt.Errorf("could not read cached data: %w", err)
+	}
+	readStart := time.Now()
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("could not read cached data: %w", err)
+	}
+	if err := json.Unmarshal(data, &cache); err != nil {
+		return nil, fmt.Errorf("could not unmarshal cached data: %w", err)
+	}
+	if err := reader.Close(); err != nil {
+		return nil, fmt.Errorf("could not read cached data: %w", err)
+	}
+	logger.Infof("Loaded %d distributions for %d identifiers after %s.", len(cache.Data), len(cache.DataByMetaData), time.Since(readStart))
+	return &cache, nil
+}
+
+// StoreCache prunes and stores cached query data to the canonical location in GCS.
+func StoreCache(storer Storer, metricName string, data *CachedQuery, logger *logrus.Entry) error {
+	pruneStart := time.Now()
+	logger.Debug("Pruning cached Prometheus data.")
+	data.prune()
+	logger.Debugf("Pruned cached Prometheus data after %s.", time.Since(pruneStart))
+
+	flushStart := time.Now()
+	logger.Info("Flushing Prometheus data to cache.")
+	raw, err := json.Marshal(&data)
+	if err != nil {
+		return fmt.Errorf("could not marshal cached data: %w", err)
+	}
+	writer, err := storer.Store(metricName + ".json")
+	if err != nil {
+		return fmt.Errorf("could open cache for writing: %w", err)
+	}
+	if _, err := writer.Write(raw); err != nil {
+		return fmt.Errorf("could not write cached data: %w", err)
+	}
+	if err := writer.Close(); err != nil {
+		return fmt.Errorf("could not write cached data: %w", err)
+	}
+	logger.Infof("Flushed Prometheus data to cache after %s.", time.Since(flushStart))
```

Comment:
Since this is going to run in a server we should use proper structured logging instead of `printf` IMHO

Reply:
I agree - but durations should be metrics, not logs anyway. I at least use structured log fields for filtering, so e.g. `org` and `repo` are useful in Prow but not `duration`. I expect to extend this thing to expose all these durations as counters in Prom for inspection.

Reply:
No one is going to filter for a duration but they are going to filter for this log message. Which with this implementation requires regex/glob filtering.